zerver/actions/message_flags.py

```diff
@@ -250,18 +251,33 @@ def do_update_message_flags(
 
     # And then create historical UserMessage records.  See the called function for more context.
     create_historical_user_messages(user_id=user_profile.id, message_ids=historical_message_ids)
-
-    if operation == "add":
-        count = msgs.update(flags=F("flags").bitor(flagattr))
-    elif operation == "remove":
-        count = msgs.update(flags=F("flags").bitand(~flagattr))
-
+    with transaction.atomic():
+        if operation == "add":
+            msgs = (
+                msgs.select_for_update()
+                .order_by("message_id")
+                .extra(where=[UserMessage.where_flag_is_absent(flagattr)])
+            )
+            updated_message_ids = [um.message_id for um in msgs]
+            msgs.filter(message_id__in=updated_message_ids).update(flags=F("flags").bitor(flagattr))
+        elif operation == "remove":
+            msgs = (
+                msgs.select_for_update()
+                .order_by("message_id")
+                .extra(where=[UserMessage.where_flag_is_present(flagattr)])
+            )
+            updated_message_ids = [um.message_id for um in msgs]
+            msgs.filter(message_id__in=updated_message_ids).update(
+                flags=F("flags").bitand(~flagattr)
+            )
+
+    count = len(updated_message_ids)
```

Comment:
I think after this line, we should have an `if count == 0: return 0, []` check; all the below code is pointless if no messages were updated.

Reply:
Makes sense!